### PR TITLE
Support filtering in backtrace command

### LINKED ISSUE #137: Support filtering frames in backtrace command
Current `backtrace` always prints all frames, which could be a huge number in Rails applications (in my project it could be 170 frames). Some simple filtering would be helpful:

1. `backtrace [num]` - only prints the first num frames.
2. `backtrace /controller/` - only prints frames with paths that match `controller`

### PR DESCRIPTION
**Pull request by st0012**
Closes #137

### CODE REVIEW

File: lib/debug/session.rb [hunk 1]
```diff
@@ -428,8 +428,24 @@ def process_command line
 
       # * `bt` or `backtrace`
       #   * Show backtrace (frame) information.
+      # * `bt <num>`
```
Comment (ko1) [inline lib/debug/session.rb:431]: `bt <num>` or `backtrace <num>` ?

File: lib/debug/session.rb [hunk 2]
```diff
@@ -428,8 +428,24 @@ def process_command line
...
+      #   * Only shows frames with location info that matches `/location_regexp/`.
+      # * `bt <num> /location_regexp/`
+      #   * Only shows first `<num>` frames with location info that matches `/location_regexp/`.
       when 'bt', 'backtrace'
-        @tc << [:show, :backtrace]
+        case arg
+        when /\A(\d+)\z/
+          @tc << [:show, :backtrace, arg.to_i]
+        when /\A\/.*\/\z/
+          @tc << [:show, :backtrace, eval(arg)]
```
Comment (ko1) [inline lib/debug/session.rb:442]: do not use `eval`, but use `Regexp.compile`.

Comment (ko1) [reply lib/debug/session.rb:442]: The message format should be `[:show, :backtrace, num, filter_regexp]`.
Do not mix the types.

Comment (ko1) [reply lib/debug/session.rb:442]: and simply `nil` represents "not specified".

File: lib/debug/thread_client.rb [hunk 3]
```diff
@@ -359,21 +359,27 @@ def frame_eval src, re_raise: false
...
+            @target_frames.select { |f| f.location_str.match?(location_pattern) }
+          else
+            @target_frames
+          end
+
+        size = frames.size
         max.times{|i|
           break if i >= size
-          puts frame_str(i)
+          puts frame_str(i, frames: frames)
```
Comment (ko1) [inline lib/debug/thread_client.rb:382]: Before filter

```
#0: foo
#1: bar
#2: boo
#3: baz
```

with filter /a/

```
#0: bar
#1: baz
```

But it is another option:

```
#1: bar
#3: baz
````

(remain original index).

Pros:
* We can recognize how deep frames.
Cons:
* Need to input huge number like 123 even if only 4 frames are shown.

Comment (ko1) [reply lib/debug/thread_client.rb:382]: I prefer remaining original index. But not sure how users feel...

Comment (st0012) [reply lib/debug/thread_client.rb:382]: I think it provides better information to users 👍 
But this requires assigning the frame index to `FrameInfo`, which is a bit outside the scope of this PR. Can I open another one for this?

Comment (ko1) [reply lib/debug/thread_client.rb:382]: You don't need to store the index into FrameInfo, but you need to collect with index like,

```ruby
fs = []
frames.each_with_index([]){|f, i|
  r << [i, f] if f.match(pat)
}```

Comment (ko1) [reply lib/debug/thread_client.rb:382]: ah, and we can choose this frame with `frame` command with original index.

Comment (st0012) [reply lib/debug/thread_client.rb:382]: I see 🤔 I'll open another PR for this after this one is merged.

File: lib/debug/thread_client.rb [hunk 4]
```diff
@@ -549,7 +555,19 @@ def wait_next_action
 
           case type
           when :backtrace
-            show_frames
+            max_lines, pattern = args
```
Comment (ko1) [inline lib/debug/thread_client.rb:558]: This code can make more shorter. I'll fix it.

**Comment by ko1** [comment]
c9d9f58910d0f6eb2f641996730bbe12a99e57a9 fixed the index issue.
Also support to filter by method name.
Do you have any problem about that?

**Comment by st0012** [comment]
of course I don't have a problem about that 😅 
I think it should be helpful too 👍

**Pull request merged**